Ignore block without transaction

riemannulus opened the pull request:
Blocks without transactions are not accepted to prevent selfish mining.

dahlia commented on Lib9c/BlockPolicy.cs:
```diff
@@ -162,6 +162,16 @@ private InvalidBlockException ValidateMinerAuthority(Block<NCAction> block)
+            if (block.Transactions.Count <= 0)
+            {
+                return new InvalidMinerException(
+                    $"The block #{block.Index} {block.Hash}'s miner {miner} should be proven by " +
+                    "including a least one of transaction.",
```
As it is intended to validate blocks, not miners…

```suggestion
                    $"The block #{block.Index} {block.Hash} (mined by {miner}) should " +
                    "include a least one of transaction.",
```

area363 replied:
Could you also change it to "at least one transaction"? :pray:

riemannulus replied:
https://github.com/planetarium/lib9c/pull/462/commits/3c20bc68d25860ab74eba48e20fb1983d77c29b2 I fixed it

riemannulus commented:
PTAL @dahlia @area363

longfin commented on Lib9c/BlockPolicy.cs:
```diff
@@ -162,6 +162,16 @@ private InvalidBlockException ValidateMinerAuthority(Block<NCAction> block)
                 );
             }
+            
+            // To prevent selfish mining, we define a consensus that blocks with no transactions are do not accepted. 
+            if (block.Transactions.Count <= 0)
+            {
+                return new InvalidMinerException(
+                    $"The block #{block.Index} {block.Hash} (mined by {miner}) should " +
+                    "include at least one transaction.",
+                    miner
+                );
+            }
```
Since the 0tx problem is about block, not miner, I think it is right to treat it as a different layer from the auth-miner.

How about checking in the `ValidateNextBlock()` instead?

riemannulus replied:
https://github.com/planetarium/lib9c/pull/462/commits/0e33021ec0110fce4406215a3c5d0b5556a6673a accept your idea, I implement `ValidateBlock` and move this check to `ValidateBlock` from `ValidateMinerAuthority`

kfangw reviewed (approved):
LGTM

area363 commented:
Test fails at https://github.com/planetarium/lib9c/pull/462/checks?check_run_id=2837870520#step:6:1717 and https://github.com/planetarium/lib9c/pull/462/checks?check_run_id=2837870520#step:6:1774 😢

riemannulus commented:
> Test fails at https://github.com/planetarium/lib9c/pull/462/checks?check_run_id=2837870520#step:6:1717 and https://github.com/planetarium/lib9c/pull/462/checks?check_run_id=2837870520#step:6:1774 😢

Yeah, I fixed this problem at 3d62b49. some tests are mined with 0tx so I added the transaction when block is mined
